### PR TITLE
Unboxed numbers

### PR DESCRIPTION
**Pull request by Ekdohibs**
This adds boxing and unboxing primitives for all boxed integers, as well as floats. This pull request does not however include frontend access to these, as it is unsafe: for this, see the pull request #1166.

### CODE REVIEW

**Comment by mshinwell** [comment]
I'll review this

File: middle_end/flambda2/from_lambda/lambda_to_flambda_primitives.ml [hunk 1]
```diff
@@ -722,6 +722,35 @@ let convert_lprim ~big_endian (prim : L.primitive) (args : Simple.t list)
...
+  | Punbox_float, [arg] ->
+    Unary (Unbox_number Flambda_kind.Boxable_number.Naked_float, arg)
+  | Pbox_float mode, [arg] ->
+    Unary
+      ( Box_number
+          ( Flambda_kind.Boxable_number.Naked_float,
+            Alloc_mode.For_allocations.from_lambda mode ~current_region ),
+        arg )
+  | Punbox_int bi, [arg] ->
+    let kind =
```
[the remark this answers is not in the document]
Comment (mshinwell) [reply middle_end/flambda2/from_lambda/lambda_to_flambda_primitives.ml:734]: That said, why don't you move this conversion into `Flambda_kind.Boxable_number`?  It's duplicated below at the moment.

File: middle_end/flambda/flambda_to_clambda.ml [hunk 2]
```diff
@@ -752,7 +754,13 @@ and to_clambda_closed_set_of_closures t env symbol
...
     let body =
       let body, body_layout = to_clambda t env_body function_decl.body in
-      assert(Lambda.compatible_layout body_layout function_decl.return_layout);
+      if not (Lambda.compatible_layout body_layout function_decl.return_layout) then begin
+        Format.printf "body: %a@.function: %a@.%a@."
+          Printlambda.layout body_layout
+          Printlambda.layout function_decl.return_layout
+          Printclambda.clambda body;
+        Misc.fatal_error "Incomatible layouts"
+      end;
```
Comment (mshinwell) [inline middle_end/flambda/flambda_to_clambda.ml:763]: Typo, and you can just use `Misc.fatal_errorf`.

File: middle_end/clambda_primitives.ml [hunk 3]
```diff
@@ -171,4 +177,7 @@ and raise_kind = Lambda.raise_kind =
 
 let equal (x: primitive) (y: primitive) = x = y
 
-let result_layout _p = Lambda.layout_any_value
+let result_layout (p : primitive) =
+  match p with
+  | Punbox_float -> Lambda.Punboxed_float
```
Comment (mshinwell) [inline middle_end/clambda_primitives.ml:182]: This is missing a case for unboxing ints, no?

**Review by mshinwell — CHANGES_REQUESTED** [review]

File: middle_end/flambda2/from_lambda/lambda_to_flambda_primitives.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (Ekdohibs) [reply middle_end/flambda2/from_lambda/lambda_to_flambda_primitives.ml:734]: Actually the function already existed in the file, using it now.

File: backend/cmmgen.ml [hunk 4]
```diff
@@ -1305,6 +1318,22 @@ and transl_let env str (layout : Lambda.layout) id exp transl_body =
...
+      let cexp = transl env exp in
+      let cbody = transl_body env in
+      match str with
+      | (Immutable | Immutable_unique) ->
+        Clet(id, cexp, cbody)
+      | Mutable ->
+        let typ = match layout with
+          | Punboxed_float -> typ_float
+          | Punboxed_int (Pint32 | Pnativeint) -> typ_int
+          | Punboxed_int Pint64 -> typ_int64
```
[the remark this answers is not in the document]
Comment (Ekdohibs) [reply backend/cmmgen.ml:1331]: I think both are the same on 64-bit architectures, but it seemed safer this way.

File: middle_end/flambda/flambda_to_clambda.ml [hunk 5]
```diff
@@ -754,13 +754,11 @@ and to_clambda_closed_set_of_closures t env symbol
     in
     let body =
       let body, body_layout = to_clambda t env_body function_decl.body in
-      if not (Lambda.compatible_layout body_layout function_decl.return_layout) then begin
-        Format.printf "body: %a@.function: %a@.%a@."
+      if not (Lambda.compatible_layout body_layout function_decl.return_layout) then
+        Misc.fatal_errorf "Imcompatible layouts:@.body: %a@.function: %a@.%a@."
```
Comment (mshinwell) [inline middle_end/flambda/flambda_to_clambda.ml:758]: Imcompatible -> Incompatible

**Review by mshinwell — CHANGES_REQUESTED** [review]

File: middle_end/flambda/flambda_to_clambda.ml (hunk 5 above)
Comment (Ekdohibs) [reply middle_end/flambda/flambda_to_clambda.ml:758]:
```suggestion
        Misc.fatal_errorf "Incompatible layouts:@.body: %a@.function: %a@.%a@."
```

**Review by mshinwell — APPROVED** [review]

**Pull request merged**